[Regression 303218@main] Follow-up fix

geoffreygaren opened the pull request:
#### 8cfcca38d4377f1a7f2ef24d62246d73f72c6550
<pre>
[Regression 303218@main] Follow-up fix
<a href="https://bugs.webkit.org/show_bug.cgi?id=302913">https://bugs.webkit.org/show_bug.cgi?id=302913</a>
<a href="https://rdar.apple.com/165178613">rdar://165178613</a>

Reviewed by Mike Wyrzykowski.

Fixed two more tiny races when loading from weak.

Also made sure not to load from m_cachedSamplerState while initializing it.

* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::tryCacheSamplerState const):

Canonical link: <a href="https://commits.webkit.org/303374@main">https://commits.webkit.org/303374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/654662eee31dad33aab18efe21e084ade90dad8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139715 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4642 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4454 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101050 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d94c677f-e713-42a4-8d58-7c46e712f010) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135146 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118409 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81845 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3430d2bf-6c91-480f-b4e0-b6e05cf2638f) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82935 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142362 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4362 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37108 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109428 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4443 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109609 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3302 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114681 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57608 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20545 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4416 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4248 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67862 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4375 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -324,18 +324,16 @@ static void uncacheSamplerState(const Sampler::UniqueSamplerIdentifier& samplerI
 
 id<MTLSamplerState> Sampler::tryCacheSamplerState() const
 {
-    if (m_cachedSamplerState)
```
isn't removing this problematic because `BindGroup.mm:1234` does this:
```
id<MTLSamplerState> sampler = apiSampler->tryCacheSamplerState();
```

should it be (in BindGroup.mm:1234):
```
id<MTLSamplerState> sampler = apiSampler->cachedSamplerState() ?: apiSampler->tryCacheSamplerState();
```
similar to what occurs in `BindGroup::rebindSamplersIfNeeded()`

geoffreygaren replied:
I guess I'll leave the existing structure to make this diff clearer.

mwyrzykowski reviewed (approved):
other than the comment looks good - thank you 👍

geoffreygaren committed:
[Regression 303218@main] Follow-up fix
https://bugs.webkit.org/show_bug.cgi?id=302913
rdar://165178613

Reviewed by Mike Wyrzykowski.

Fixed two more tiny races when loading from weak.

Also made sure not to load from m_cachedSamplerState while initializing it.

* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::tryCacheSamplerState const):

Canonical link: https://commits.webkit.org/303374@main
